main.go

```diff
@@ -102,46 +119,72 @@ func main() {
 		})
 	}
 
+	if len(orphanFlags.Contents) > 0 {
+		parallelApply(orphanFlags, func(o types.Object) {
+			if err := cleanObject(ctx, s3Cli, o); err != nil {
+				log.Error().Err(err).Msg("Cleaning error")
+			}
+		})
+	}
+
 	log.Info().
 		Dur("elapsed", time.Since(start)).
 		Msg("Processing ended !")
 }
 
-func filterBucketContent(elems s3.ListObjectsOutput, csvSuffix, processedFlagSuffix string) (unprocessed, processed s3.ListObjectsOutput) {
+func filterBucketContent(elems s3.ListObjectsOutput, csvSuffix, processedFlagSuffix string) (unprocessed, processed, orphanFlags s3.ListObjectsOutput) {
 	// Rely on .processed files present on the bucket to detect which csv
 	// has already been pushed to influx and which has yet to be processed
+	// List .processed files that do not match any data file in order to clean them up, this can happen if the crawler was interrupted
 	csvFiles := []types.Object{}
-	processedElems := []string{}
+	flags := []types.Object{}
+	objectNames := []string{}
+	flagNames := []string{}
 
 	if len(csvSuffix) == 0 {
-		return unprocessed, processed
+		return unprocessed, processed, orphanFlags
 	}
 	for _, s := range elems.Contents {
 		if strings.HasSuffix(*s.Key, csvSuffix) {
 			csvFiles = append(csvFiles, s)
+			objectNames = append(objectNames, *s.Key)
 		}
 		if strings.HasSuffix(*s.Key, processedFlagSuffix) {
-			processedElems = append(processedElems, strings.ReplaceAll(*s.Key, processedFlagSuffix, csvSuffix))
+			flags = append(flags, s)
+			flagNames = append(flagNames, *s.Key)
 		}
 	}
 
 	for _, o := range csvFiles {
-		if !slices.Contains(processedElems, *o.Key) {
+		flagName := strings.ReplaceAll(*o.Key, csvSuffix, processedFlagSuffix)
+		if !slices.Contains(flagNames, flagName) {
 			unprocessed.Contents = append(unprocessed.Contents, o)
 		} else {
 			processed.Contents = append(processed.Contents, o)
 		}
 	}
 
-	return unprocessed, processed
+	for _, f := range flags {
+		fileName := strings.ReplaceAll(*f.Key, processedFlagSuffix, csvSuffix)
+		if !slices.Contains(objectNames, fileName) {
+			orphanFlags.Contents = append(orphanFlags.Contents, f)
+		}
+
+	}
+
+	return unprocessed, processed, orphanFlags
 }
 
 func parallelApply(list s3.ListObjectsOutput, fn func(o types.Object)) {
+	//Limit the number of parallel routines doing the processing.
+	semaphore := make(chan struct{}, opts.MaxRoutines)
 	var wg sync.WaitGroup
 	wg.Add(len(list.Contents))
 	for _, item := range list.Contents {
 		o := item
 		go func() {
+			semaphore <- struct{}{}        // Lock channel
+			defer func() { <-semaphore }() // Release channel
```

Comment:
First time I see a channel used as a semaphore ! 
As this code complexify, I wonder if https://pkg.go.dev/golang.org/x/sync/errgroup could help (as you want).

Reply:
Thanks for the suggestion, this indeed looks simpler !